Add <input type=checkbox switch> macOS vertical rendering

annevk opened the pull request:
#### 24ebb01cef269ea5bfd77c2abc82d56474206a77
<pre>
Add &lt;input type=checkbox switch&gt; macOS vertical rendering
<a href="https://bugs.webkit.org/show_bug.cgi?id=266347">https://bugs.webkit.org/show_bug.cgi?id=266347</a>

Reviewed by Aditya Keerthi.

This adjusts the logic in CheckboxInputType so pointer tracking is
based on the vertical position when the control is rendered
vertically.

It similarly adjusts the macOS drawing logic, with some additional
special casing for the on label which shows when
&quot;Differentiate without color&quot; is enabled.

Rather than continuing to reuse
adjustButtonOrCheckboxOrColorWellOrInnerSpinButtonOrRadioOrSwitchStyle()
we make switch break from that pattern by selectively copying over the
parts that apply to switch. (Spoiler: setting the logical width and
height.)

And finally, but not least, adjust the repaint size calculation.
<a href="https://bugs.webkit.org/show_bug.cgi?id=266532">https://bugs.webkit.org/show_bug.cgi?id=266532</a> is filed as a follow-up
for this due to an issue with writing-mode:vertical-rl.

* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/checkbox-switch-input-computed-style.tentative-expected.txt:
* Source/WebCore/html/CheckboxInputType.cpp:
(WebCore::switchPointerTrackingLogicalLeftPosition):
(WebCore::CheckboxInputType::startSwitchPointerTracking):
(WebCore::CheckboxInputType::stopSwitchPointerTracking):
(WebCore::CheckboxInputType::isSwitchPointerTracking const):
(WebCore::CheckboxInputType::updateIsSwitchVisuallyOnFromAbsoluteLocation):
* Source/WebCore/html/CheckboxInputType.h:
* Source/WebCore/platform/Theme.h:
(WebCore::Theme::inflateControlPaintRect const): Deleted.
* Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.h:
* Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.mm:
(WebCore::SwitchMacUtilities::visualCellSize):
(WebCore::SwitchMacUtilities::visualCellOutsets):
(WebCore::SwitchMacUtilities::rectWithTransposedSize):
(WebCore::SwitchMacUtilities::trackRectForBounds):
(WebCore::SwitchMacUtilities::rotateContextForVerticalWritingMode):
* Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.mm:
(WebCore::SwitchThumbMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm:
(WebCore::trackMaskImage):
(WebCore::trackImage):
(WebCore::SwitchTrackMac::draw):
* Source/WebCore/platform/mac/ThemeMac.h:
* Source/WebCore/platform/mac/ThemeMac.mm:
(WebCore::visualSwitchMargins):
(WebCore::ThemeMac::inflateControlPaintRect const):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::adjustButtonOrCheckboxOrColorWellOrInnerSpinButtonOrRadioStyle const):
(WebCore::RenderTheme::adjustCheckboxStyle const):
(WebCore::RenderTheme::adjustRadioStyle const):
(WebCore::RenderTheme::adjustColorWellStyle const):
(WebCore::RenderTheme::adjustButtonStyle const):
(WebCore::RenderTheme::adjustInnerSpinButtonStyle const):
(WebCore::RenderTheme::adjustSwitchStyle const):
(WebCore::RenderTheme::adjustButtonOrCheckboxOrColorWellOrInnerSpinButtonOrRadioOrSwitchStyle const): Deleted.
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::adjustSwitchStyle const):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::adjustRepaintRect):

Canonical link: <a href="https://commits.webkit.org/272405@main">https://commits.webkit.org/272405@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5918ae4a8917ffd5263a33b0b908d51fc47867ee

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31519 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10194 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33224 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34008 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/28545 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32290 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12545 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7442 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28159 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31860 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8584 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28128 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7388 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7553 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28036 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35351 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28646 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28483 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/33682 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7634 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5650 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31528 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9291 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7405 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8322 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8138 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on LayoutTests/fast/forms/switch/size-expected.txt:
```diff
@@ -0,0 +1,20 @@
+layer at (0,0) size 800x600
```
We generally prefer to avoid adding new render tree dump test cases, as they are difficult to maintain. This test really just needs to compare inline and block sizes.

See the `computed-style` tests in `LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms` for examples.

pxlcoder commented on Source/WebCore/html/CheckboxInputType.cpp:
```diff
@@ -231,14 +231,21 @@ void CheckboxInputType::didDispatchClick(Event& event, const InputElementClickSt
     event.setDefaultHandled();
 }
 
+static int switchPointerTrackingXOrYPosition(Element& element, LayoutPoint absoluteLocation)
```
`logicalX` or `logicalLeft` instead of `XOrYPosition`.

pxlcoder commented on Source/WebCore/html/CheckboxInputType.cpp:
```diff
@@ -392,7 +399,7 @@ bool CheckboxInputType::isSwitchVisuallyOn() const
 void CheckboxInputType::updateIsSwitchVisuallyOnFromAbsoluteLocation(LayoutPoint absoluteLocation)
 {
-    auto xPosition = element()->renderer()->absoluteToLocal(absoluteLocation, UseTransforms).x();
+    auto xOrYPosition = switchPointerTrackingXOrYPosition(*element(), absoluteLocation);
```
`logicalX`/`logicalLeft`

pxlcoder commented on Source/WebCore/html/CheckboxInputType.cpp:
```diff
@@ -392,7 +399,7 @@ bool CheckboxInputType::isSwitchVisuallyOn() const
+    auto xOrYPosition = switchPointerTrackingXOrYPosition(*element(), absoluteLocation);
     auto isSwitchVisuallyOn = m_isSwitchVisuallyOn;
     auto isRTL = element()->computedStyle()->direction() == TextDirection::RTL;
     auto switchThumbIsLeft = (!isRTL && !isSwitchVisuallyOn) || (isRTL && isSwitchVisuallyOn);
```
`switchThumbIsLogicallyLeft`?

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.mm:
```diff
@@ -82,6 +82,26 @@ static float easeInOut(const float progress)
+    return outsets.top() * style.zoomFactor - outsets.bottom() * style.zoomFactor;
+}
+
+static void adjustContextForVertical(GraphicsContext& context, const FloatPoint origin, float x)
```
`rotateContextForVerticalWritingMode`

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm:
```diff
@@ -116,10 +116,14 @@
         (__bridge NSString *)kCUIScaleKey: @(deviceScaleFactor),
     }];
 
+    // This ensures the on label continues to appear upright. It's not pretty but it works.
```
`It's not pretty but it works.` is an unnecessary comment.

pxlcoder commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1524,9 +1524,18 @@ void RenderTheme::adjustSliderThumbStyle(RenderStyle& style, const Element* elem
+    auto controlSize = Theme::singleton().controlSize(StyleAppearance::Switch, style.fontCascade(), zoomedSize, style.effectiveZoom());
+
+    if (style.isHorizontalWritingMode()) {
+        style.setMinWidth(WTFMove(controlSize.width));
```
`setLogicalMinWidth` and `setLogicalMinHeight`, so you don't need an `if` here.

pxlcoder commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1524,9 +1524,18 @@ void RenderTheme::adjustSliderThumbStyle(RenderStyle& style, const Element* elem
+void RenderTheme::adjustSwitchStyle(RenderStyle& style, const Element*) const
 {
-    adjustButtonOrCheckboxOrColorWellOrInnerSpinButtonOrRadioOrSwitchStyle(style, element);
+    auto zoomedSize = style.isHorizontalWritingMode() ? LengthSize { style.width(), style.height() } : LengthSize { style.height(), style.width() };
```
This looks wrong. What if `style.width()` or `style.height()` are not fixed values.

Need to check `.isFixed()`. Also we have `logicalWidth()` and `logicalHeight()` to avoid checking `style.isHorizontalWritingMode()` here.

annevk replied:
`switchSize()` in `ThemeMac` handles this.

annevk replied:
I double checked the code in `adjustButtonOrCheckboxOrColorWellOrInnerSpinButtonOrRadioOrSwitchStyle()` btw and I feel more confident now that this is identical to what it did before (except now accounting for "logical" and with much less no-op code).

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.mm:
```diff
@@ -82,6 +82,26 @@ static float easeInOut(const float progress)
+static float verticalInflatedRectHeightOffset(const IntOutsets& outsets, const ControlStyle& style)
+{
+    return outsets.top() * style.zoomFactor - outsets.bottom() * style.zoomFactor;
+}
```
This might be unnecessary if you transpose outsets. See the vertical writing mode logic in `ProgressBarMac::rectForBounds`.

annevk replied:
I don't understand how you can adjust the outsets before you draw. Don't they have to be correct for horizontal drawing?

pxlcoder replied:
The outsets simply determine the inflated rect, which is always done before drawing. There wouldn't be any change to horizontal mode, since the outsets would only be adjusted for vertical writing mode. 

That's what `ProgressBarMac` does in `rectForBounds` under `if (isVerticalWritingMode)`. 

Currently, the switch drawing logic calls `inflatedRect` directly. For example, `auto inflatedTrackRect = inflatedRect(trackRect, size, outsets, style);`. But I think it's time to implement `rectForBounds` to return an inflated rect that you can use to avoid additional adjustments later on.

Then, once you have that set up, similar to `ProgressBarMac::draw`, again under `if (isVerticalWritingMode)`, you'd transpose the the size of the inflated rect.

pxlcoder reviewed (commented):
I think the WPT import should be done separately. Especially, since it touches files unrelated to this change.

The reason being, if this change needs to be reverted for any reason, it doesn't make sense to lose test coverage too.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.mm:
```diff
@@ -49,7 +49,22 @@ static IntSize cellSize(NSControlSize controlSize)
+
+static IntOutsets logicalCellOutsets(NSControlSize controlSize, bool isVertical)
 {
     static const IntOutsets margins[] =
```
Can we keep this in `cellOutsets` and call that method here?

annevk replied:
Yeah we could, it just wasn't clear what utility that would provide. There's no code calling into cellOutsets from a super class.

pxlcoder replied:
I'd eventually want to call it to avoid the duplicate cell outset definitions used in the Web process side. Since we already have it implemented, and the code itself can still be used in this method, I would keep it.

annevk replied:
I see, I thought we'd use some kind of statics file for that which both sides include. Anyway, happy to make it work again.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.mm:
```diff
@@ -82,6 +100,28 @@ static float easeInOut(const float progress)
+    return { FloatPoint { bounds.x(), bounds.y() + offsetY }, size };
+}
+
+static void rotateContextForVerticalWritingMode(GraphicsContext& context, const FloatPoint origin, float x, float y)
```
The `x` parameter here is unnecessary, since it's always going to be the height. And the `y` parameter is specific to a single adjustment (on/off) labels, that does not need to be generalized.

This could just be `rotateContextForLogicalRect(GraphicsContext& context, FloatRect& rect)`. And any additional translation you need (like the on/off) label can be made inline.

Eventually, this could just be a method somewhere like `ControlMac`, so `MeterMac` and `ProgressBarMac` could also use it. That would be for another PR though.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm:
```diff
@@ -76,7 +76,7 @@
 }
 
-static RefPtr<ImageBuffer> trackImage(GraphicsContext& context, RefPtr<ImageBuffer> trackMaskImage, FloatSize trackRectSize, float deviceScaleFactor, const ControlStyle& style, bool isOn, bool isRTL, bool isEnabled, bool isPressed, bool isInActiveWindow, bool needsOnOffLabels, NSString * coreUISize)
+static RefPtr<ImageBuffer> trackImage(GraphicsContext& context, RefPtr<ImageBuffer> trackMaskImage, FloatSize trackRectSize, float deviceScaleFactor, const ControlStyle& style, bool isOn, bool isRTL, bool isVertical, bool isEnabled, bool isPressed, bool isInActiveWindow, bool needsOnOffLabels, NSString * coreUISize)
```
Nit, that I just noticed:
```suggestion
static RefPtr<ImageBuffer> trackImage(GraphicsContext& context, RefPtr<ImageBuffer> trackMaskImage, FloatSize trackRectSize, float deviceScaleFactor, const ControlStyle& style, bool isOn, bool isRTL, bool isVertical, bool isEnabled, bool isPressed, bool isInActiveWindow, bool needsOnOffLabels, NSString *coreUISize)
```

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm:
```diff
@@ -185,17 +192,23 @@
     if (isFocused) {
         auto color = colorFromCocoaColor([NSColor keyboardFocusIndicatorColor]).opaqueColor();
-        context.drawFocusRing(Vector { trackRect }, 0, 1.0f, color);
+        context.drawFocusRing(Vector { logicalTrackRect }, 0, 0, color);
```
Isn't `logicalTrackRect` really a physical rect, given the lack of rotation needed here?

annevk commented on Source/WebCore/platform/mac/ThemeMac.mm:
```diff
@@ -652,11 +662,15 @@ static LengthSize switchSize(const LengthSize& zoomedSize, float zoomFactor)
+        if (isVertical)
+            zoomedSize = zoomedSize.transposedSize();
+        zoomedRect = inflateRect(zoomedRect, zoomedSize, logicalSwitchMargins(controlSize, isVertical), zoomFactor);
         break;
```
There is something wrong with this logic as not everything is redrawn for a 10x10 vertical switch. In particular there's a bar on the right side that remains static.

(This code might also need further tidying in general, but I wanted to put it up to maybe get another review cycle before the weekend.)

annevk replied:
As far as I can tell this is doing the correct thing. Something with repaint seems amiss.

annevk replied:
Minimal test:
`data:text/html,<input%20type=checkbox%20switch%20style=width:10px;height:10px;writing-mode:vertical-rl>`

annevk replied:
I filed https://bugs.webkit.org/show_bug.cgi?id=266532 on this. I'm quite positive now this code is correct. It produces equivalent results for horizontal and vertical. It just goes amiss with vertical due to the repaint code.

annevk replied:
(The one thing I have now revised here is renamed `logicalSwitchMargins` to `visualSwitchMargins`. This is still correct otherwise.)

pxlcoder replied:
Talked about this offline, none of the paint inflation code for controls accounts for flipped blocks writing mode, so we will fix separately.

annevk committed:
Add <input type=checkbox switch> macOS vertical rendering
https://bugs.webkit.org/show_bug.cgi?id=266347

Reviewed by Aditya Keerthi.

This adjusts the logic in CheckboxInputType so pointer tracking is
based on the vertical position when the control is rendered
vertically.

It similarly adjusts the macOS drawing logic, with some additional
special casing for the on label which shows when
"Differentiate without color" is enabled.

Rather than continuing to reuse
adjustButtonOrCheckboxOrColorWellOrInnerSpinButtonOrRadioOrSwitchStyle()
we make switch break from that pattern by selectively copying over the
parts that apply to switch. (Spoiler: setting the logical width and
height.)

And finally, but not least, adjust the repaint size calculation.
https://bugs.webkit.org/show_bug.cgi?id=266532 is filed as a follow-up
for this due to an issue with writing-mode:vertical-rl.

* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/checkbox-switch-input-computed-style.tentative-expected.txt:
* Source/WebCore/html/CheckboxInputType.cpp:
(WebCore::switchPointerTrackingLogicalLeftPosition):
(WebCore::CheckboxInputType::startSwitchPointerTracking):
(WebCore::CheckboxInputType::stopSwitchPointerTracking):
(WebCore::CheckboxInputType::isSwitchPointerTracking const):
(WebCore::CheckboxInputType::updateIsSwitchVisuallyOnFromAbsoluteLocation):
* Source/WebCore/html/CheckboxInputType.h:
* Source/WebCore/platform/Theme.h:
(WebCore::Theme::inflateControlPaintRect const): Deleted.
* Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.h:
* Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.mm:
(WebCore::SwitchMacUtilities::visualCellSize):
(WebCore::SwitchMacUtilities::visualCellOutsets):
(WebCore::SwitchMacUtilities::rectWithTransposedSize):
(WebCore::SwitchMacUtilities::trackRectForBounds):
(WebCore::SwitchMacUtilities::rotateContextForVerticalWritingMode):
* Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.mm:
(WebCore::SwitchThumbMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm:
(WebCore::trackMaskImage):
(WebCore::trackImage):
(WebCore::SwitchTrackMac::draw):
* Source/WebCore/platform/mac/ThemeMac.h:
* Source/WebCore/platform/mac/ThemeMac.mm:
(WebCore::visualSwitchMargins):
(WebCore::ThemeMac::inflateControlPaintRect const):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::adjustButtonOrCheckboxOrColorWellOrInnerSpinButtonOrRadioStyle const):
(WebCore::RenderTheme::adjustCheckboxStyle const):
(WebCore::RenderTheme::adjustRadioStyle const):
(WebCore::RenderTheme::adjustColorWellStyle const):
(WebCore::RenderTheme::adjustButtonStyle const):
(WebCore::RenderTheme::adjustInnerSpinButtonStyle const):
(WebCore::RenderTheme::adjustSwitchStyle const):
(WebCore::RenderTheme::adjustButtonOrCheckboxOrColorWellOrInnerSpinButtonOrRadioOrSwitchStyle const): Deleted.
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::adjustSwitchStyle const):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::adjustRepaintRect):

Canonical link: https://commits.webkit.org/272405@main
